RFC: Provider default style overrides (#23853)
microsoft/fluentui: rfcs/react-components/convergence/provider-style-overrides.md

# RFC: Provider style overrides

@ling1726

## Summary

> Proposes two methods of overriding the default styles on the `FluentProvider`

Reviewer: ```suggestion
Proposes two methods of overriding the default styles on the `FluentProvider`.
```

## Problem statement

The `FluentProvider` ships with [default styles](https://github.com/microsoft/fluentui/blob/656f002f3062c0a7962fb23eb165dec64ecc1509/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderStyles.ts#L11-L18) which can be too opinionated for some customers, especially those who have
strict requirements on global styles in their application.

> Currently it's not possible to use `makeStyles` to override these default styles. See the example below:

Reviewer: It's not entirely true, I suggest to change it:

```suggestion
If `makeStyles` will be used to override these default styles it will result in unexpected issues until additional wrapping will be done. See the example below:
```

```tsx
import { makeStyles } from '@fluentui/react-components';

const useProviderOverrides = makeStyles({
  provider: {
    // ⚠️ requires marginRight RTL style
    marginLeft: '1px',
  },
});

const App = () => {
  const styles = useProviderOverrides();
  return <FluentProvider className={styles.provider}>{children}</FluentProvider>;
};
```

> The above example is provlematic because the `TextDirectionProvider` is rendered

Author: we would still need `TextDirectionProvider` because there's no guarantee that folks would use logical properties. If we did decide to enforce this then.... I guess we're going full on v10 :)

> The above example is problematic because the `TextDirectionProvider` is rendered
> internally inside the `FluentProvider`. The provider is required to let `makeStyles` know to resolve and apply RTL
> styles to DOM.

Reviewer: ```suggestion
`FluentProvider` renders internally `TextDirectionProvider` (from `@griffel/react`). The `TextDirectionProvider` is required to let `makeStyles` know to when to return RTL classes.

In the example above `makeStyles` is called *outside* `FluentProvider` (and `TextDirectionProvider`) thus will have a wrong value for text direction. This problem is not specific to Fluent UI React and can be faced in other React applications.
```

IMO it sounds more accurate.

While it is possible to override these styles using global CSS in a stylesheet by targeting the provider, that would
not handle portals which use the same default styles as the closest `FluentProvider` ancestor.

## Detailed Design or Proposal

### Document usage of extra providers

We can document how to override `FluentProvider` styles using the required extra providers so that the
default provider styles can be overriden.

```tsx
import { makeStyles, RendererProvider, TextDirectionProvider } from '@fluentui/react-components';

const useProviderOverrides = makeStyles({
  provider: {
    marginLeft: '1px',
  },
});

const App = ({ dir }) => {
  const styles = useProviderOverrides();
  return (
    <FluentProvider dir={dir} className={styles.provider}>
      {children}
    </FluentProvider>
  );
};

const AppWrapper = () => {
  return (
    <TextDirectionProvider dir="rtl">
      <App dir="rtl" />
    </TextDirectionProvider>
  );
};
```

#### Pros and Cons

> 👍 No extra code changes necessary
> 👍 Users will understand what the problem is
> 👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
> 
> 👎 Extra effort to override one single component

Author: ```suggestion
- 👍 No extra code changes necessary
 -👍 Users will understand what the problem is
 -👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)

- 👎 Extra effort to override one single component
```

> - 👍 No extra code changes necessary
>  -👍 Users will understand what the problem is
>  -👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
> 
> - 👎 Extra effort to override one single component

Author: ```suggestion
- 👍 No extra code changes necessary
- 👍 Users will understand what the problem is
- 👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)

- 👎 Extra effort to override one single component
```

> - 👎 Extra effort to override one single component

Reviewer: Not sure if it's a real problem as it will be done in a limited amount of places

Author: I would say it's still extra effort because it's not obvious and even with documentation it's hard to know to specifically look for it either

> 

Reviewer: ```suggestion
- 👍/👎 `TextDirectionProvider` should be re-exported

```

We can extend this method to customers by creating a new API `makeProviderStyles` which would leverage `@griffel/core`
internally. `makeProviderStyles` would return a `makeStyles` factory that should be passed to the `FluentProvider`
and applied in the same way as the default provider styles.

```ts
// makeProviderStyles.ts

import { makeStyles, GriffelStyle } from '@griffel/core';

const makeProviderStyles = (providerStyles: GriffelStyle) => {
  return makeStyles({
    providerStyles,
  });
};
```

By leveraging `makeProviderStyles` the consuming app can override its styles without needing to know much else
about how the `FluentProvider` or Griffel work internally.

```tsx
// App.tsx
import { makeProviderStyles } from '@fluentui/react-components';

const providerStyles = makeProviderStyles({
  marginLeft: '1px',
});

const App = () => {
  return <FluentProvider styles={providerStyles}>{children}</FluentProvider>;
};
```

#### Pros and Cons

- 👍 Add styles the same way as the `FluentProvider` currently does it
- 👍 No need to know about provider internals
- 👍 One time thing to do for `FluentProvider` only if really necessary

> - 👎 slightly different API to `makeStyles`

Reviewer: ```suggestion
- 👎 Slightly different API to `makeStyles`
- 👎 `className` prop will be still exposed so users would be still able to pass classes from `makeStyles()` call
```

This is also a bummer as then there is no a single way of doing overrides

> 👍 Add styles the same way as the `FluentProvider` currently does it
> 👍 No need to know about provider internals
> 👍 One time thing to do for `FluenProvider` only if really necessary
> 
> 👎 slightly different API to `makeStyles`
> 👎 something 'special' to do for one component

Author: ```suggestion
- 👍 Add styles the same way as the `FluentProvider` currently does it
- 👍 No need to know about provider internals
- 👍 One time thing to do for `FluenProvider` only if really necessary

- 👎 slightly different API to `makeStyles`
- 👎 something 'special' to do for one component
```

> - 👎 something 'special' to do for one component

Reviewer: ```suggestion
- 👎 Something 'special' to do for one component i.e. one time API
- 👎 Additional handling for build time transforms
```

# RFC: Provider style overrides

@ling1726

## Summary

Proposes two methods of overriding the default styles on the `FluentProvider`.

## Problem statement

The `FluentProvider` ships with [default styles](https://github.com/microsoft/fluentui/blob/656f002f3062c0a7962fb23eb165dec64ecc1509/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderStyles.ts#L11-L18) which can be too opinionated for some customers, especially those who have
strict requirements on global styles in their application.

> If the Provider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:

Reviewer: ```suggestion
If the FluentProvider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```

> If the FluentProvider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:

Reviewer 2: ```suggestion
If the FluentProvider is styled using a `makeStyles` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```

We can extend this method to customers by creating a new API `makeProviderStyles` which would leverage `@griffel/core`
internally. `makeProviderStyles` would return a `makeStyles` factory that should be passed to the `FluentProvider`
and applied in the same way as the default provider styles.

```ts
// makeProviderStyles.ts

import { makeStyles, GriffelStyle } from '@griffel/core';

const makeProviderStyles = (providerStyles: GriffelStyle) => {
  return makeStyles({
    providerStyles,
  });
};
```

By leveraging `makeProviderStyles` the consuming app can override its styles without needing to know much else
about how the `FluentProvider` or Griffel work internally.

```tsx
// App.tsx
import { makeProviderStyles } from '@fluentui/react-components';

const providerStyles = makeProviderStyles({
  marginLeft: '1px',
});

const App = () => {
  return <FluentProvider styles={providerStyles}>{children}</FluentProvider>;
};
```

> Additionaly in the `FluentProvider` we can strip and classnames that are known to be created by a standard `makeStyles`

Reviewer 2: ```suggestion
Additionaly in the `FluentProvider` we will strip and classnames that are known to be created by a standard `makeStyles`
```
